lib/integrations/samlidp/gcpworkforce.go

```diff
@@ -0,0 +1,215 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package samlidp
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log/slog"
+	"net/http"
+	"time"
+
+	"github.com/gravitational/trace"
+	"google.golang.org/api/googleapi"
+	iam "google.golang.org/api/iam/v1"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/lib/integrations/samlidp/samlidpconfig"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// GCPWorkforceService defines workforce service configuration
+// parameters.
+type GCPWorkforceService struct {
+	// APIParams holds basic input params required to create GCP workforce
+	// pool and pool provider.
+	APIParams samlidpconfig.GCPWorkforceAPIParams
+	// HTTPClient is used to fetch metadata from the SAMLIdPMetadataURL
+	// endpoint.
+	HTTPClient *http.Client
+}
+
+// NewGCPWorkforceService creates a new GCPWorkforceService.
+func NewGCPWorkforceService(cfg GCPWorkforceService) (*GCPWorkforceService, error) {
+	newGCPWorkforceService := &GCPWorkforceService{
+		APIParams: samlidpconfig.GCPWorkforceAPIParams{
+			PoolName:           cfg.APIParams.PoolName,
+			PoolProviderName:   cfg.APIParams.PoolProviderName,
+			OrganizationID:     cfg.APIParams.OrganizationID,
+			SAMLIdPMetadataURL: cfg.APIParams.SAMLIdPMetadataURL,
+		},
+		HTTPClient: cfg.HTTPClient,
+	}
+
+	if err := newGCPWorkforceService.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return newGCPWorkforceService, nil
+}
+
+// CreateWorkforcePoolAndProvider creates GCP Workforce Identity Federation pool and pool provider
+// with the given GCPWorkforceAPIParams values.
+func (s *GCPWorkforceService) CreateWorkforcePoolAndProvider(ctx context.Context) error {
+	slog.InfoContext(ctx, "Configuring Workforce Identity Federation pool and SAML provider.")
+
+	iamService, err := iam.NewService(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	workforceService := iam.NewLocationsWorkforcePoolsService(iamService)
+
+	slog.With("pool name", s.APIParams.PoolName).InfoContext(ctx, "Creating workforce pool.")
```

Comment:
I think we should not use spaces in log fields
The rfd is not explicit about it, but the used examples seem to indicate so.
https://github.com/gravitational/teleport/blob/master/rfd/0154-logging-guidelines.md#withfields

Reply:
used `_` in spaces. looking at codebase its a mixed bag of cameCase and underscore but I think underscore is better for log readibility? https://github.com/gravitational/teleport/pull/39820/commits/d662a0b837c3a1ceb6224008ce391b2090668a72